expat/lib/xmlparse.c

```diff
@@ -1954,22 +1954,16 @@ XML_Parse(XML_Parser parser, const char *s, int len, int isFinal) {
                       &parser->m_position);
     nLeftOver = s + len - end;
     if (nLeftOver) {
-      if (parser->m_buffer == NULL
-          || nLeftOver > parser->m_bufferLim - parser->m_buffer) {
-        /* avoid _signed_ integer overflow */
-        char *temp = NULL;
-        const int bytesToAllocate = (int)((unsigned)len * 2U);
-        if (bytesToAllocate > 0) {
-          temp = (char *)REALLOC(parser, parser->m_buffer, bytesToAllocate);
-        }
-        if (temp == NULL) {
-          parser->m_errorCode = XML_ERROR_NO_MEMORY;
-          parser->m_eventPtr = parser->m_eventEndPtr = NULL;
-          parser->m_processor = errorProcessor;
-          return XML_STATUS_ERROR;
-        }
-        parser->m_buffer = temp;
-        parser->m_bufferLim = parser->m_buffer + bytesToAllocate;
+      // hacky way to avoid ERROR_SUSPENDED from XML_GetBuffer
+      const enum XML_Parsing status = parser->m_parsingStatus.parsing;
+      parser->m_parsingStatus.parsing = XML_PARSING;
+      char *const temp = XML_GetBuffer(parser, len + 1);
```

Comment:
The old code used `len * 2` for some reason (strange, since `len` may have been much bigger than `nLeftOver`). I used `len + 1` on the assumption that `len` would be a power of two, and the +1 would push us to the next power of two in buffer size (guaranteeing that both `nLeftOver` and the next `len` fill from the app will fit, without realloc).

Maybe I should just use `nLeftOver` for simplicity.

Reply:
Done in upcoming push.